tla-io/src/main/scala/at/forsyte/apalache/io/tnt/TntIR.scala

```diff
@@ -0,0 +1,346 @@
+      scala.util.Failure(new TntIRParseError(err.clue))
+    }
+}
+
+case class TntModule(
+    id: Int,
+    name: String,
+    defs: List[TntDef])
+object TntModule {
+  implicit val rw: RW[TntModule] = macroRW
+}
+
+/** The representation of types in the type map */
+case class TntTypeScheme(
+    @key("type") typ: TntType
+    // TODO Will we need these for anything?
+    // typeVariables: List[String],
+    // rowVariables: List[String]
+  )
+object TntTypeScheme {
+  implicit val rw: RW[TntTypeScheme] = macroRW
+}
+
+/** Source IDs, used to associate expressions with their inferred types */
+trait WithID {
+  val id: Int
+}
+
+/** The representation of TNT expressions */
+sealed trait TntEx extends WithID
+
+object TntEx {
+
+  implicit val rw: RW[TntEx] =
+    RW.merge(Name.rw, Bool.rw, TntInt.rw, Str.rw, App.rw, Lambda.rw, Let.rw)
+
+  /** A name of: a variable, constant, parameter, user-defined operator */
+  @key("name") case class Name(id: Int, name: String) extends TntEx {}
+  object Name {
+    implicit val rw: RW[Name] = macroRW
+  }
+
+  /** The boolean literal value */
+  @key("bool") case class Bool(id: Int, value: Boolean) extends TntEx {}
+  object Bool {
+    implicit val rw: RW[Bool] = macroRW
+  }
+
+  @key("int") case class TntInt(id: Int, value: Int) extends TntEx {}
+  object TntInt {
+    implicit val rw: RW[TntInt] = macroRW
+  }
+
+  @key("str") case class Str(id: Int, value: String) extends TntEx {}
+  object Str {
+    implicit val rw: RW[Str] = macroRW
+  }
+
+  @key("app") case class App(
+      id: Int,
+      /** The name of the operator being applied */
+      opcode: String,
+      /** A list of arguments to the operator */
+      args: List[TntEx])
+      extends TntEx {}
+  object App {
+    implicit val rw: RW[App] = macroRW
+  }
+
+  @key("lambda") case class Lambda(
+      id: Int,
+      /** Identifiers for the formal parameters */
+      params: List[String],
+      /** The qualifier for the defined operator */
+      // TODO should this eventually be a sumtype?
+      qualifier: String,
+      /** The definition body */
+      expr: TntEx)
+      extends TntEx {}
+  object Lambda {
+    implicit val rw: RW[Lambda] = macroRW
+  }
+
+  @key("let") case class Let(
+      id: Int,
+      /** The operator being defined for use in the body */
+      opdef: TntDef.Def,
+      /** The body */
+      expr: TntEx)
+      extends TntEx {}
+  object Let {
+    implicit val rw: RW[Let] = macroRW
+  }
+}
+
+/** TNT Definitions (e.g., of types, operators, constants, etc.) */
+sealed trait TntDef extends WithID
+
+object TntDef {
+  implicit val rw: RW[TntDef] = RW.merge(Def.rw, Var.rw, Const.rw, Assume.rw, TypeDef.rw)
+
+  trait WithOptionalTypeAnnotation {
+    val typeAnnotation: Option[TntType]
+  }
+
+  trait WithTypeAnnotation {
+    val typeAnnotation: TntType
+  }
+
+  /**
+   * A user-defined operator that is defined via one of the qualifiers: val, def, pred, action, or temporal. Note that
```

Comment:
```suggestion
   * A user-defined operator that is defined via one of the qualifiers: val, def, action, or temporal. Note that
```

It was probably copied from an outdated comment.

Reply:
Also, while we don't use this qualifier information (which can be forever afaiu), I think we should keep that as a string and even get rid of the value options in the comments since that will be out of sync pretty easily.